frontend/server/handlers/artifacts.ts

```diff
@@ -239,6 +254,111 @@ function getGCSArtifactHandler(options: { key: string; bucket: string }, peek: n
   };
 }
 
+function getVolumeArtifactsHandler(options: { bucket: string; key: string }, peek: number = 0) {
+  const { key, bucket } = options;
+  return async (req: Request, res: Response) => {
+    let filePath;
+    let volumeMount;
+    try {
+      // get ml-pipeline-ui pod namespace
+      const SERVER_NAMESPACE = serverInfo.getServerNamespace();
+      // get ml-pipeline-ui pod name
+      const POD_NAME = serverInfo.getServerPodName();
+
+      // only check when POD_NAME and serverNamespace can be obtained
+      if (POD_NAME && SERVER_NAMESPACE) {
+        const [pod, err] = await k8sHelper.getPod(POD_NAME, SERVER_NAMESPACE);
+        if (err) {
+          const { message, additionalInfo } = err;
+          console.error(message, additionalInfo);
+          res.status(500).send(message);
+          return;
+        }
+
+        if (!pod) {
+          const message = `Could not get pod ${POD_NAME} in namespace ${SERVER_NAMESPACE}`;
+          res.status(500).send(message);
+          return;
+        }
+
+        // volumes not specified or volume named ${bucket} not specified
+        if (!Array.isArray(pod?.spec?.volumes) || !pod.spec.volumes.find(v => v?.name === bucket)) {
+          const message = `Failed to open volume://${bucket}/${key}, volume ${bucket} not exist`;
+          res.status(400).send(message);
+          return;
+        }
+
+        // find volumes mount
+        if (
+          Array.isArray(pod?.spec?.containers) &&
+          Array.isArray(pod.spec.containers[0]?.volumeMounts)
+        ) {
+          volumeMount = pod.spec.containers[0].volumeMounts.find(v => {
+            // volume name must be same
+            if (v?.name === bucket) {
+              if (v?.subPath) {
+                return key.startsWith(v.subPath);
+              } else {
+                return true;
+              }
+            }
+            return false;
+          });
+        }
+      } else {
+        console.error(
+          `pod name or server namespace can't be obtained, pod name: ${POD_NAME}, server namespace: ${SERVER_NAMESPACE}`,
+        );
+        res
+          .status(500)
+          .send(
+            `Failed to open volume://${bucket}/${key}, can't obtain server pod name or server pod namespace`,
+          );
+        return;
+      }
+
+      // volumes mount not exist
+      if (!volumeMount) {
+        const message = `Failed to open volume://${bucket}/${key}, volume mount not exist`;
+        res.status(400).send(message);
+        return;
+      }
+
+      // relative file path
+      const relativeFilePath = volumeMount.subPath
+        ? key.substring(volumeMount.subPath.length)
+        : key;
+      filePath = path.join(volumeMount.mountPath, relativeFilePath);
```

Comment:
As discussed, the logic for finding this filePath is quite complex. Would you mind extracting it to a function and unit test it.